Additions for evolution

nilsdeppe opened the pull request:
## Proposed changes

**Note:** Either #377 or this PR will need to be checked to make sure the tensor get function change doesn't break anything.

- Fix two bugs in Wedge3D
- Allow converting between ElementId and ElementIndex
- Provide place to register CoordinateMap classes to be serializable through the base class. Needed for serializing ElementMap
- Add ElementMap, the map that goes from the Logical coordinates to the Grid coordinates in evolutions where there are time dependent maps. For time independent maps it could go straight to the Inertial frame.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."

nilsdeppe committed:
Allow creating ElementId from ElementIndex

nilsdeppe committed:
Make CoordinateMapBase from DomainCreators serializable

nilsdeppe committed:
Add ElementMap class

An ElementMap is the Block CoordinateMap prefixed with an AffineMap in
each dimension that maps to the section Element size.

nilsdeppe committed:
Fix serialization in CoordinateMaps::Wedge3D

nilsdeppe commented on src/Domain/DomainCreators/RegisterDerivedWithCharm.cpp:
```diff
@@ -0,0 +1,25 @@
+
+namespace DomainCreators {
+void register_derived_with_charm() {
+  PUPable_reg(SINGLE_ARG(::CoordinateMap<Frame::Logical, Frame::Inertial,
```
I'm thinking of how to keep this list maintainable in the long run. We should probably have functions for 1D, 2D and 3D. Alphabetical by map name, left to right within each function?

wthrowe commented on src/Domain/ElementIndex.hpp:
```diff
@@ -53,6 +57,12 @@ class ElementIndex {
   ElementIndex() = default;
   explicit ElementIndex(const ElementId<VolumeDim>& id) noexcept;
+  size_t block_id() const noexcept { return segments_[0].block_id(); }
+  const std::array<ElementIndex_detail::SegmentIndex, VolumeDim>& segments()
```
I don't like having a "_detail" class in the public interface.  Make it not-detail or something.

kidder replied:
I think I agree; I'd just remove SegmentId from the ElementIndex_detail namespace.

wthrowe commented on src/Domain/ElementMap.hpp:
```diff
@@ -0,0 +1,134 @@
+ * \brief The CoordinateMap for the Element from the Logical frame to the
+ * `TargetFrame`
+ *
+ *
```
Was something supposed to go here?

nilsdeppe replied:
Awe, RIP :( Yea, this needs more documentation...

wthrowe commented on src/Domain/ElementMap.hpp:
```diff
@@ -0,0 +1,134 @@
+                    SpatialIndex<Dim, UpLo::Lo, Frame::Logical>>>
+  jacobian(tnsr::I<T, Dim, Frame::Logical> source_point) const noexcept {
+    for (size_t d = 0; d < Dim; ++d) {
+      source_point.get(d) = source_point.get(d) * gsl::at(map_slope_, d) +
```
I think this is the third time I've seen this loop now.  Perhaps factor it into a helper function?

wthrowe commented on src/Domain/ElementMap.hpp:
```diff
@@ -0,0 +1,134 @@
+  //                       segment_ids[i].endpoint(Side::Lower)
+  std::array<double, Dim> length_of_range_{
+      make_array<Dim>(std::numeric_limits<double>::signaling_NaN())};
+  // length_of_range_[i] = segment_ids[i].endpoint(Side::Upper) +
```
It is not clear to me what these comments have to do with the code surrounding them. Additionally, this one appears to contradict the previous one.

nilsdeppe replied:
The reason I added the comments was so it was clearer what the variable below the comment corresponds to (clearly wasn't successful :) ). However, since it is the same as in the constructor I would say it's better to delete them. Do you agree?

wthrowe replied:
I kind of like the idea, but even knowing what they are supposed to do I still can't decipher them.  What is `A_`?  What is `Dim.0`?  And I think the one on this line is referring to the wrong variable.

I also wouldn't object if you'd prefer to remove them and just let people look at the constructor since the calculations aren't very complicated in this case.

wthrowe commented on src/Domain/ElementMap.cpp:
```diff
@@ -0,0 +1,67 @@
+void ElementMap<Dim, TargetFrame>::pup(PUP::er& p) noexcept {
+  p | block_map_;
+  p | element_id_;
+  p | map_slope_;
```
`length_of_range_` and `sum_of_endpoints_` are missing.  They are also unused, except during initialization.  Can we not store them?

nilsdeppe replied:
Sure. Do you think adding a function to compute them that is called in the lambdas, or moving the initialization of everything that follows to the constructor body is better?

wthrowe replied:
How about just using `2. * map_slope_` for `length_of_range_` and `2. * map_offset_` for `sum_of_endpoints_`?

wthrowe commented on src/Domain/ElementMap.hpp:
```diff
@@ -0,0 +1,134 @@
+  std::array<double, Dim> map_inverse_offset_{
+      make_array<Dim>(std::numeric_limits<double>::signaling_NaN())};
+  // length_of_range_ / length_of_domain_ = 0.5 * length_of_range_
+  std::array<double, Dim> jacobian_{
```
A brief comment that the Jacobian is diagonal and so these are the diagonal entries would make things clearer.

kidder commented on src/Domain/ElementId.hpp:
```diff
@@ -67,6 +74,10 @@ template <size_t VolumeDim>
 bool operator!=(const ElementId<VolumeDim>& lhs,
                 const ElementId<VolumeDim>& rhs) noexcept;
 
+template <size_t VolumeDim>
```
This function already exists in Domain/InitalElementIds.hpp , although it does it for all blocks and not a specific one.

nilsdeppe replied:
So what is your request? Move this function to there and have the one for all blocks call this one?

kidder replied:
Is there a reason to call it on only one block?  If no, my suggestion is to remove the function; otherwise, yes move the function

nilsdeppe replied:
I use it on a single Block in some of the other evolution things, at least right now. I've added an overload to InitialElementIds and the ones for all blocks call the ones for single blocks. I've also separated this change into a separate commit.

kidder commented on src/Domain/ElementIndex.hpp:
```diff
@@ -29,6 +29,10 @@ class SegmentIndex {
+  size_t block_id() const noexcept { return block_id_; }
+  size_t index() const noexcept { return index_; }
+  size_t refinement_level() const noexcept { return refinement_level_; }
+
```
with these defined you can use them in the stream operator and it no longer needs to be a friend

kidder commented on tests/Unit/Domain/DomainCreators/Test_Brick.cpp:
```diff
@@ -167,6 +168,24 @@ SPECTRE_TEST_CASE("Unit.Domain.DomainCreators.Brick", "[Domain][Unit]") {
+          AffineMap3D{AffineMap{-1., 1., lower_bound[0], upper_bound[0]},
+                      AffineMap{-1., 1., lower_bound[1], upper_bound[1]},
+                      AffineMap{-1., 1., lower_bound[2], upper_bound[2]}});
+  const auto base_map_deserialized = serialize_and_deserialize(base_map);
```
In order to test the serialization, shouldn't we be checking that base_map_deserialized == base_map?

kidder commented on tests/Unit/Domain/DomainCreators/Test_Brick.cpp:
```diff
@@ -167,6 +168,24 @@ SPECTRE_TEST_CASE("Unit.Domain.DomainCreators.Brick", "[Domain][Unit]") {
+      AffineMap3D{AffineMap{-1., 1., lower_bound[0], upper_bound[0]},
+                  AffineMap{-1., 1., lower_bound[1], upper_bound[1]},
+                  AffineMap{-1., 1., lower_bound[2], upper_bound[2]}});
+  CHECK(*dynamic_cast<MapType>(base_map.get()) == coord_map);
```
can we simplify this by calling the are_maps_equal function in CoordMaps?  And are we testing anything in addition to what is tested by the check_if_maps_are_equal in test_domain_construction?

wthrowe commented on src/Domain/ElementId.hpp:
```diff
@@ -29,12 +33,15 @@ class ElementId {
 
+  /// Convert an ElementIndex to an ElementId
+  // clang-tidy: mark explicit: we want to allow conversion
+  ElementId(const ElementIndex<VolumeDim>& index) noexcept;  // NOLINT
```
What is the intended use for this as an implicit conversion?

nilsdeppe replied:
Since a function inside an Action might take an ElementId, but the passed in index in the Charm++ array an implicit conversion would simplify calling such functions. Do you think that's a bad idea?

nilsdeppe committed:
Add function to compute initial ElementId for single block

kidder commented on src/Domain/InitialElementIds.cpp:
```diff
@@ -3,24 +3,74 @@
-      element_ids.emplace_back(
-          block_id, make_array<1>(SegmentId(xi_refinement_level, xi_index)));
-    }
+    auto ids_for_block =
```
can combine into a single template and explicit instantiate by dim

kidder reviewed (changes requested):
squash

nilsdeppe commented:
Made a few updates:

1. Rebased on develop since merges
2. Fixed clang-tidy warning about operator<< redundant declaration for ElementId
3. Have Wedge3D test of ElementMap check the inverse, since that was added
4. Have ElementMap tests check that the block_map() function works correctly.
